File: day-05/binary_boarding.py
input_file = 'day-05/input.txt'

trans = str.maketrans('FBLR', '0101')
def get_seat_id(boarding_string):
  return int(boarding_string.translate(trans), 2)

def part1(input):
  seat_ids = sorted([get_seat_id(boarding_string) for boarding_string in input.splitlines()])
  return max(seat_ids)

def part2(input):
  seat_ids = sorted([get_seat_id(boarding_string) for boarding_string in input.splitlines()])
  for i in seat_ids:
    if ((i + 1) not in seat_ids) and ((i + 2) in seat_ids):
      return i + 1

if __name__ == "__main__":
  with open(input_file) as f:
    data = f.read()
  print("Part 1: ", part1(data))
  print("Part 2: ", part2(data))

File: day-05/binary_boarding.go
package main

import (
	"fmt"
	"io/ioutil"
	"sort"
	"strconv"
	"strings"
)

const input_file string = "./day-05/input.txt"

func ParseInput(input string) []int {
	seat_ids := []int{}
	replacer := strings.NewReplacer("F", "0", "B", "1", "L", "0", "R", "1")
	lines := strings.Split(strings.TrimSpace(input), "\n")
	for _, line := range lines {
		seat_id, _ := strconv.ParseInt(replacer.Replace(line), 2, 0)
		seat_ids = append(seat_ids, int(seat_id))
	}
	sort.Ints(seat_ids)
	return seat_ids
}

func Part1(input string) int {
	seat_ids := ParseInput(input)
	return seat_ids[len(seat_ids)-1]
}

func Part2(input string) int {
	seat_ids := ParseInput(input)
	for i, value := range seat_ids {
		if seat_ids[i+1] != value+1 {
			return value + 1
		}
	}
	return 0
}

func main() {
	b, err := ioutil.ReadFile(input_file)
	if err != nil {
		fmt.Print(err)
	}
	data := string(b)

	fmt.Printf("Part 1: %d\n", Part1(data))
	fmt.Printf("Part 2: %d\n", Part2(data))
}
